features/support/env.rb

```diff
@@ -167,21 +167,40 @@ def synchronize(seconds = session_options.default_max_wait_time, options = {})
       # Note this is done *after* the possible patching via synchronize_with_timeout_error, so that it wraps
       # *outside* that.
 
-      def synchronize_with_unload_wait(*args, &block)
+      def is_page_loading()
+        browser_params = driver.evaluate_script(%q({
+          "navigationStart": window.performance.timing.navigationStart,
+          "loadEventEnd": window.performance.timing.loadEventEnd,
+          "xhr_semaphore": (window.GOVUK && window.GOVUK.xhr_semaphore) || 0
+        }))
+
+        if browser_params['navigationStart'] > browser_params['loadEventEnd']
+          # navigationStart has been updated more recently than loadEventEnd, loadEventEnd presumably still
+          # carrying the value set when the *old* page got loaded - that means the dom is probably in the
+          # process of loading (or at least requesting) a new page. any following page queries will
+          # presumably be targeted at the upcoming page, which isn't ready.
+          reason = "navigationStart (#{browser_params['navigationStart']}) > loadEventEnd (#{browser_params['loadEventEnd']})"
+        elsif browser_params['xhr_semaphore'] > 0
+          # window.GOVUK.xhr_semaphore's non-zero value indicates that the page still has an ajax request in flight
+          reason = "xhr_semaphore = #{browser_params['xhr_semaphore']}"
+        else
+          return false
+        end
+
+        DEBUG_FILE.write "#{Time.now.utc.round(10).iso8601(6)}: #{reason} @ #{caller.to_a.select { |pth| pth.include? '/features/' }}\n"
```

Comment:
Oooh - where does `DEBUG_FILE` go then? Can we see it in Jenkins somewhere?

Reply:
You _would_ be able to, except it has an unpredictable filename because it includes the pid. However if you ssh in you should see at least one `debug.<n>.out` in the same directory as the rest of the reports.